Tests/ApolloInternalTestHelpers/MockURLProtocol.swift

```diff
@@ -0,0 +1,56 @@
+import Foundation
+
+public class MockURLProtocol<RequestProvider: MockRequestProvider>: URLProtocol {
+  
+  override class public func canInit(with request: URLRequest) -> Bool {
+    return true
+  }
+  
+  override class public func canonicalRequest(for request: URLRequest) -> URLRequest {
+    return request
+  }
+  
+  override public func startLoading() {
+    guard let url = self.request.url,
+          let handler = RequestProvider.requestHandlers[url] else {
+      fatalError("No URL available for URLRequest.")
+    }
+    
+    DispatchQueue.main.asyncAfter(deadline: .now() + Double.random(in: 0.0...0.5)) {
```

Comment:
Thoughts on the time delay range? Is 0.5 seconds to high in the range for testing purposes? Wondering if I should close this range down even further

Reply:
I _think_ it's fine. Since it was using a service before we have no idea what the real delay was but this seems like a wide enough variance.

Reply:
I do think that 0.5 seconds is pretty long. I'd probably do `0.0...0.2` personally. But it's definitely not an issue either way.

Reply:
My one concern here is if we should be dispatching this on the `main` queue. Does `URLSessionClient` traditionally dispatch to the main thread? My assumption is that it would call on a background thread. And we should be trying to mimic the real life behaviors in these tests as closely as possible.

Reply:
We [default to the main queue](https://github.com/apollographql/apollo-ios/blob/main/Sources/Apollo/URLSessionClient.swift#L65) but that can be overridden.

Reply:
Ahhh yea looks like the delegates and completion handlers are default to the main queue. So what do we think about having this changed to the global queue vs staying on main?

Reply:
It will definitely change during 2.0 but I'm skeptical whether we should make the change now. It's arguably better behaviour but we have no idea what calls people making out of the callback. It could be disastrous for the less informed.

Reply:
You should use `DispatchQueue(label: "MockURLProtocol", qos: .background)`.